packages/sw-cache-expiration/src/lib/behavior.js

```diff
@@ -14,144 +14,257 @@
       now = Date.now();
     }
 
-    return this.db.then((db) => {
-      const tx = db.transaction(this.cacheName, 'readwrite');
-      tx.objectStore(this.cacheName).put(now, url);
-      return tx.complete;
+    const db = await this.getDB();
+    const tx = db.transaction(this.cacheName, 'readwrite');
+    tx.objectStore(this.cacheName).put({
+      [timestampProperty]: now,
+      [urlProperty]: url,
     });
+
+    await tx.complete;
   }
 
   /**
-   * TODO expireEntries description
-   * @param {Number} now Defaults to the current time
-   * @return {Promise} List of removed entries.
+   * Expires entries, both based on the the maximum age and the maximum number
+   * of entries, depending on how this instance is configured.
+   *
+   * @param {Object} input The input object to this function.
+   * @param {Number} [input.now] A timestamp. Defaults to the current time.
+   * @return {Array<string>} A list of the URLs that were expired.
    */
-  expireEntries(now) {
+  async expireEntries({now} = {}) {
     if (typeof now === 'undefined') {
       now = Date.now();
     }
 
-    const promises = [];
-    promises.push(
-      this.maxAgeSeconds ? this._expireOldEntries(now) : Promise.resolve([]));
-    promises.push(
-      this.maxEntries ? this._expireExtraEntries() : Promise.resolve([]));
+    // First, expire old entries, if maxAgeSeconds is set.
+    const oldEntries = this.maxAgeSeconds ?
+      await this.findOldEntries({now}) :
+      [];
+
+    // Once that's done, check for the maximum size.
+    const extraEntries = this.maxEntries ?
+      await this.findExtraEntries() :
+      [];
+
+    const urls = [...new Set(oldEntries.concat(extraEntries))];
+    await this.deleteFromCacheAndIDB({urls});
 
-    return Promise.all(promises)
-      .then(([oldEntries, extraEntries]) => oldEntries.concat(extraEntries));
+    return urls;
   }
 
   /**
+   * Expires entries based on the the maximum age.
+   *
    * @private
-   * @param {Number} now
-   * @return {Promise<Array<String>>} Promise that resolves to an
-   *         array of urls.
+   * @param {Object} input The input object to this function.
+   * @param {Number} [input.now] A timestamp.
+   * @return {Array<string>} A list of the URLs that were expired.
    */
-  _expireOldEntries(now) {
-    if (typeof now === 'undefined') {
-      now = Date.now();
-    }
+  async findOldEntries({now} = {}) {
+    assert.isType({now}, 'number');
 
     const expireOlderThan = now - (this.maxAgeSeconds * 1000);
     const urls = [];
-    return this.db.then((db) => {
-      const tx = db.transaction(this.cacheName, 'readwrite');
-      const store = tx.objectStore(this.cacheName);
-      store.iterateCursor((cursor) => {
+    const db = await this.getDB();
+    const tx = db.transaction(this.cacheName, 'readonly');
+    const store = tx.objectStore(this.cacheName);
+    const index = store.index(timestampProperty);
+
+    index.iterateCursor(async (cursor) => {
```

Comment:
Oops, this was just leftover from a previous version in which I was doing `IndexedDB` operations inside the loop. I'll remove it.